### PR TITLE
EndingBlankLines checker

### PR DESCRIPTION
**Pull request by evgeniy-r**
I have added the required check.
I check only the last one character and it seems that it is enough (however, it is easy to add more sequences of characters).

#### ✔ Checklist:
- [x] This PR has been added to [CHANGELOG.md](https://github.com/dotenv-linter/dotenv-linter/blob/master/CHANGELOG.md) (at the top of the list);
- [x] Tests for the changes have been added (for bug fixes / features);
- [x] Docs have been added / updated (for bug fixes / features).

### CODE REVIEW

**Comment by mgrachev** [comment]
@evgeniy-r Thank you for your contribution! 🚀

I need more time to review your PR.

**Comment by evgeniy-r** [comment]
I have examined the issue one more time, and now I think that can solve this with less number of changes.
I think that we can just check that the last element of the `lines` collection in `checks::run` is blank.
I can update PR, if it is a good idea.

**Comment by mgrachev** [comment]
@evgeniy-r If you can do it, then please do.

**Comment by evgeniy-r** [comment]
It is ready, take a look, please.

**Comment by evgeniy-r** [comment]
I have rebase it on the updated master branch.

File: src/lib.rs [hunk 1]
```diff
@@ -38,6 +37,46 @@ fn get_args(current_dir: &OsStr) -> clap::ArgMatches {
...
 
+fn run_checks(fe: &FileEntry) -> io::Result<Vec<Warning>> {
+    let f = File::open(fe.path.clone())?;
+    let mut reader = BufReader::new(f);
+
+    // TODO: Initialize a vector with a capacity equal to the number of lines
+    let mut lines: Vec<LineEntry> = Vec::new();
+
+    let mut number = 0;
+    loop {
```
Comment (mgrachev) [inline src/lib.rs:48]: I don't like the idea of using `loop` to extract data from a file.
Previous way with `for (index, line) in reader.lines().enumerate() {}` worked fine:
```rust
// TODO: Initialize a vector with a capacity equal to the number of lines
let mut lines: Vec<LineEntry> = Vec::new();
let mut number = 0;
for (index, line) in reader.lines().enumerate() {
    number = index + 1;
    let raw_string = line?;

    lines.push(LineEntry {
        number,
        file_path: file.path.clone(),
        raw_string,
    })
}

let result = checks::run(lines);
warnings.extend(result);
```

You can get the last character from the file more easily:
```rust
// TODO: Initialize a vector with a capacity equal to the number of lines
let mut lines: Vec<LineEntry> = Vec::new();
let mut number = 0;
for (index, line) in reader.lines().enumerate() {
// ...
}

f.seek(SeekFrom::End(-1));
let mut reader = BufReader::new(&f);
let mut s = String::new();
reader.read_to_string(&mut s)?;

lines.push(LineEntry {
    number: number + 1,
    file_path: file.path.clone(),
    raw_string: s,
});

let result = checks::run(lines);
warnings.extend(result);
```

File: src/checks.rs [hunk 2]
```diff
@@ -28,22 +29,34 @@ fn checklist() -> Vec<Box<dyn Check>> {
...
+        warnings.push(warning);
+    }
+}
+
 pub fn run(lines: Vec<LineEntry>) -> Vec<Warning> {
     let mut checks = checklist();
     let mut warnings: Vec<Warning> = Vec::new();
 
-    for line in lines {
+    for line in &lines {
```
Comment (mgrachev) [inline src/checks.rs:46]: If you need to check the last line separately, you can do something like this:
```rust
if let Some((last_line, rest_lines)) = lines.split_last() {
    for line in rest_lines {
        if line.is_empty_or_comment() {
            continue;
        }

        for ch in &mut checks {
            if let Some(warning) = ch.run(&line) {
                warnings.push(warning);
            }
        }
    }

    let mut ending_blank_line_check = ending_blank_line::EndingBlankLineChecker::default();
    if let Some(warning) = ending_blank_line_check.run(&last_line) {
        warnings.push(warning);
    }
}
```

File: src/checks.rs [hunk 3]
```diff
@@ -28,22 +29,34 @@ fn checklist() -> Vec<Box<dyn Check>> {
     ]
 }
 
+fn ending_line_checker() -> Box<dyn Check> {
+    Box::new(ending_blank_line::EndingBlankLineChecker::default())
```
Comment (mgrachev) [inline src/checks.rs:33]: Using a `Box` and a separate function for one check is not necessary.

File: src/checks.rs [hunk 4]
```diff
@@ -28,22 +29,34 @@ fn checklist() -> Vec<Box<dyn Check>> {
...
+}
+
+fn check_and_push_warning(checker: &mut dyn Check, line: &LineEntry, warnings: &mut Vec<Warning>) {
+    if let Some(warning) = checker.run(line) {
+        warnings.push(warning);
```
Comment (mgrachev) [inline src/checks.rs:38]: Moving this code to a separate function is not necessary.

File: src/checks/ending_blank_line.rs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+pub(crate) struct EndingBlankLineChecker {
+    name: &'static str,
+    template: &'static str,
+}
+
+impl Default for EndingBlankLineChecker {
+    fn default() -> Self {
+        Self {
+            name: "EndingBlankLine",
+            template: "No blank line at the end of file",
```
Comment (mgrachev) [inline src/checks/ending_blank_line.rs:13]:
```suggestion
            template: "No blank line at the end of the file",
```

File: src/checks/ending_blank_line.rs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+impl EndingBlankLineChecker {
+    fn message(&self) -> String {
+        format!("{}: {}", self.name, self.template)
+    }
+}
+
+impl Check for EndingBlankLineChecker {
+    fn run(&mut self, line: &LineEntry) -> Option<Warning> {
+        if line.is_empty() {
```
Comment (mgrachev) [inline src/checks/ending_blank_line.rs:26]: I would like to check here if the string has ending blank line. For example:
```rust
if line.raw_string.chars().any(|c| c == '\n') {
    return None;
}
```

File: src/checks/key_without_value.rs [hunk 7]
```diff
@@ -24,7 +24,10 @@ impl KeyWithoutValueChecker<'_> {
 impl Check for KeyWithoutValueChecker<'_> {
     fn run(&mut self, line: &LineEntry) -> Option<Warning> {
         if !line.raw_string.contains('=') {
-            Some(Warning::new(line.clone(), self.message(&line.raw_string)))
+            Some(Warning::new(
+                line.clone(),
+                self.message(&line.trimmed_string()),
```
Comment (mgrachev) [inline src/checks/key_without_value.rs:29]: Why are these changes needed?

File: src/common.rs [hunk 8]
```diff
@@ -63,18 +63,24 @@ pub struct LineEntry {
...
 
     pub fn get_key(&self) -> Option<String> {
         if self.is_empty_or_comment() {
             return None;
         }
 
-        match self.raw_string.find('=') {
-            Some(index) => Some(self.raw_string[..index].to_owned()),
+        match self.trimmed_string().find('=') {
+            Some(index) => Some(self.trimmed_string()[..index].to_owned()),
```
Comment (mgrachev) [inline src/common.rs:83]: Why are these changes needed?

File: src/lib.rs (hunk 1 above)
Comment (evgeniy-r) [reply src/lib.rs:48]: Yes, I tryed this approach too, I have it in the separate branch locally.
I can easily move to it.

File: src/checks.rs (hunk 2 above)
Comment (evgeniy-r) [reply src/checks.rs:46]: It is better, thank you.

File: src/common.rs (hunk 8 above)
Comment (evgeniy-r) [reply src/common.rs:83]: Because if we have this line (it is started with whitespaces) in .env file
```
    ABC=def
```
the key is
```
ABC
```
not
```
    ABC
```

Am I right? If not, this change is wrong.

File: src/checks/key_without_value.rs (hunk 7 above)
Comment (evgeniy-r) [reply src/checks/key_without_value.rs:29]: Now we have the character `\ n` in the raw string.
I think this should not appear in the message.

Comment (evgeniy-r) [reply src/checks/key_without_value.rs:29]: But if we will move back to `for (index, line) in reader.lines().enumerate()`, then it is not neccessary.

File: src/checks.rs (hunk 4 above)
Comment (evgeniy-r) [reply src/checks.rs:38]: Now we have this code in 2 places, and it seems that there will be more such places if we add new checks for empty lines or comments.

What do you think?

File: src/checks.rs (hunk 3 above)
Comment (evgeniy-r) [reply src/checks.rs:33]: I wanted the functions (`ending_line_checker` and `checklist`) to be more unified.
But I will change it, if it is unneccessary.

File: src/checks/ending_blank_line.rs (hunk 6 above)
Comment (evgeniy-r) [reply src/checks/ending_blank_line.rs:26]: First, if we go back to `for (index, line) in reader.lines().enumerate()` then we will not have `\n` in` raw_string`.

Second, let me to explain my solution: 

In `checks` we work with the representation of files as `EntryLine` collections. 
And I think that this is a good model, a good abstraction. I like it.
I suppose that in this model the file containing `\n` is two entry lines, not one (and IDE shows this file as 2 lines).
Thus, I believe that the implementation of this check is more in the correct presentation of the file as a `EntryLine` collection.

I hope that I was able to clearly explain.

File: src/lib.rs [hunk 9]
```diff
@@ -107,25 +146,58 @@ pub fn run() -> Result<Vec<Warning>, Box<dyn Error>> {
...
+        file.into_temp_path()
     }
 
-    Ok(warnings)
+    #[test]
+    fn correct_files() {
+        let contents = vec![
+            "A=B\nF=BAR\nFOO=BAR\n",
+            "A=B\r\nF=BAR\r\nFOO=BAR\r\n",
+            "# comment\nABC=DEF\n",
```
Comment (mgrachev) [inline src/lib.rs:175]: I think it is more integration test, not unit. Please, move it into the `tests` directory.

File: src/lib.rs (hunk 1 above)
Comment (evgeniy-r) [reply src/lib.rs]: Done

File: src/checks.rs (hunk 2 above)
Comment (evgeniy-r) [reply src/checks.rs]: Done

File: src/checks.rs (hunk 3 above)
Comment (evgeniy-r) [reply src/checks.rs]: Done

File: src/checks/ending_blank_line.rs (hunk 5 above)
Comment (evgeniy-r) [reply src/checks/ending_blank_line.rs]: Done

File: src/checks/key_without_value.rs (hunk 7 above)
Comment (evgeniy-r) [reply src/checks/key_without_value.rs]: Done

File: src/checks/ending_blank_line.rs (hunk 6 above)
Comment (evgeniy-r) [reply src/checks/ending_blank_line.rs]: I have made the check more clear.

File: README.md [hunk 10]
```diff
@@ -28,6 +28,7 @@
 It checks `.env` files for problems that may cause the application to malfunction:
 <p>
 &nbsp;&nbsp;&nbsp;&nbsp;✅<a href="#duplicated-key">Duplicated Key</a><br />
+&nbsp;&nbsp;&nbsp;&nbsp;✅<a href="#ending-blank-line">Ending Blank Line</a><br />
```
Comment (mgrachev) [inline README.md:31]:
```suggestion
&nbsp;&nbsp;&nbsp;&nbsp;✅<a href="#ending-blank-line-will-be-available-in-v200">Ending Blank Line</a> (Will be available in <a href="https://github.com/dotenv-linter/dotenv-linter/issues/172">v2.0.0</a>)<br />
```

File: README.md [hunk 11]
```diff
@@ -184,6 +185,21 @@ FOO=BAR
 BAR=FOO
 ```
 
+### Ending Blank Line
```
Comment (mgrachev) [inline README.md:188]:
```suggestion
### Ending Blank Line (Will be available in [v2.0.0](#172))
```

File: src/checks.rs (hunk 4 above)
Comment (mgrachev) [reply src/checks.rs:34]: I think there will be no more such places, so I would not like to move this code to the separate function.

File: src/checks/ending_blank_line.rs [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+    use std::path::PathBuf;
+
+    #[test]
+    fn blank_line() {
+        let mut checker = EndingBlankLineChecker::default();
+        let line = LineEntry {
+            number: 1,
+            file_path: PathBuf::from(".env"),
+            raw_string: String::from("\n"),
```
Comment (mgrachev) [inline src/checks/ending_blank_line.rs:46]: Could you write also a test for `\r\n`?

File: src/common.rs (hunk 8 above)
Comment (mgrachev) [reply src/common.rs:83]: I guess you are right 👍

File: src/common.rs [hunk 13]
```diff
@@ -89,6 +95,10 @@ impl LineEntry {
             None => None,
         }
     }
+
+    pub fn trimmed_string(&self) -> &str {
+        self.raw_string.trim()
```
Comment (mgrachev) [inline src/common.rs:100]: Could you write a test for that function?

File: src/lib.rs [hunk 14]
```diff
@@ -38,6 +37,43 @@ fn get_args(current_dir: &OsStr) -> clap::ArgMatches {
...
+            number,
+            file_path: fe.path.clone(),
+            raw_string,
+        })
+    }
+
+    let mut last_line = String::new();
+    let ending_seq_length = 1;
+    if f.seek(SeekFrom::End(-ending_seq_length)).is_ok()
+        && f.read_to_string(&mut last_line).unwrap() == ending_seq_length as usize
```
Comment (mgrachev) [inline src/lib.rs:63]:
```suggestion
        && f.read_to_string(&mut last_line)? == ending_seq_length as usize
```

File: src/lib.rs [hunk 15]
```diff
@@ -38,6 +37,43 @@ fn get_args(current_dir: &OsStr) -> clap::ArgMatches {
         .get_matches()
 }
 
+fn run_checks(fe: &FileEntry) -> io::Result<Vec<Warning>> {
+    let mut f = File::open(fe.path.clone())?;
```
Comment (mgrachev) [inline src/lib.rs:41]: Is calling the `clone()` method necessary?

File: src/common.rs (hunk 13 above)
Comment (evgeniy-r) [reply src/common.rs:100]: Done

File: src/checks/ending_blank_line.rs (hunk 12 above)
Comment (evgeniy-r) [reply src/checks/ending_blank_line.rs:46]: Done

File: README.md (hunk 11 above)
Comment (evgeniy-r) [reply README.md]: Done

File: README.md (hunk 10 above)
Comment (evgeniy-r) [reply README.md]: Done

File: src/lib.rs (hunk 15 above)
Comment (evgeniy-r) [reply src/lib.rs]: No, I have used ref.

File: src/lib.rs (hunk 14 above)
Comment (evgeniy-r) [reply src/lib.rs]: Done

File: src/checks.rs (hunk 4 above)
Comment (evgeniy-r) [reply src/checks.rs]: Done

File: src/lib.rs (hunk 9 above)
Comment (evgeniy-r) [reply src/lib.rs:172]: In this test I wanted to test working with a single file, independently of file browsing.

So I need to make the `run_checks` function and the `common::FileEntry` type public, to make the test an integration one.

Comment (evgeniy-r) [reply src/lib.rs:172]: What do you think, @mgrachev?

Comment (mgrachev) [reply src/lib.rs:172]: I think you do something wrong if you want to make some functions public only for tests.

Comment (mgrachev) [reply src/lib.rs:172]: The `lib.rs` file shoud check only in intergration tests.

**Comment by evgeniy-r** [comment]
> Why have you done these changes?
> tests/cli_common.rs → tests/cli_common/mod.rs

Because `cli_common.rs` itself is not a test, it is a shared module. This is not a mandatory change, but it will be more clear.

https://doc.rust-lang.org/book/ch11-03-test-organization.html#submodules-in-integration-tests

File: src/lib.rs (hunk 9 above)
Comment (evgeniy-r) [reply src/lib.rs]: > The `lib.rs` file shoud check only in intergration tests.

I agree with this, but my wish may indicate that it is needed some decomposition in `lib.rs` )

I slightly change approach and move this tests to `tests` directory.

Also I noticed that we can't check a file not from the current directory (or its subdirectories):
```
:~$ dotenv-linter 1.env
1.env:3 LowercaseKey: The fOO key should be in uppercase
:~$ cd subd
:~/subd$ dotenv-linter ~/1.env
:~/subd$
```
It is because `let strip_path = match result` in `lib.rs` (`StripPrefixError`).

Is this done on purpose?

Comment (mgrachev) [reply src/lib.rs]: > > The `lib.rs` file shoud check only in intergration tests.
> 
> I agree with this, but my wish may indicate that it is needed some decomposition in `lib.rs` )

Absolutely agree. I plan to do a refactor this file a bit later and cover it with tests.

> 
> I slightly change approach and move this tests to `tests` directory.
> 
> Also I noticed that we can't check a file not from the current directory (or its subdirectories):
> 
> ```
> :~$ dotenv-linter 1.env
> 1.env:3 LowercaseKey: The fOO key should be in uppercase
> :~$ cd subd
> :~/subd$ dotenv-linter ~/1.env
> :~/subd$
> ```
> 
> It is because `let strip_path = match result` in `lib.rs` (`StripPrefixError`).
> 
> Is this done on purpose?

I know about that problem. Could you open an issue about this problem?
There we can discuss it is a real problem or not.

File: tests/sample_files.rs [hunk 16]
```diff
@@ -0,0 +1,48 @@
+use assert_cmd::assert::Assert;
+use assert_cmd::Command;
+
+use std::io::Write;
+use tempfile::{NamedTempFile, TempPath};
+
+fn create_temp_file(contents: &str) -> TempPath {
+    let mut file = NamedTempFile::new().unwrap();
```
Comment (mgrachev) [inline tests/sample_files.rs:8]: Please use structs and functions from [cli_common.rs](https://github.com/dotenv-linter/dotenv-linter/blob/master/tests/cli_common.rs).

Comment (evgeniy-r) [reply tests/sample_files.rs:8]: There are 2 reasons why I didn't use it:
 * `TestFile::new` uses the `writeln!` macro, and it appends a new line to content (https://doc.rust-lang.org/std/macro.writeln.html).
* The idea was to check the correctness (or incorrectness) of the files without providing the exact error messages (and `TestDir::test_command_fail_with_args` requires such messages).

File: src/lib.rs (hunk 9 above)
Comment (evgeniy-r) [reply src/lib.rs]: https://github.com/dotenv-linter/dotenv-linter/issues/179

File: tests/sample_files.rs (hunk 16 above)
Comment (evgeniy-r) [reply tests/sample_files.rs:8]: Can I make some changes in `cli_common`?

Also, I can leave only the tests that relate to checking the blank line at the end of the file (and rename the test), if you don't like this approach (arrays of the right and wrong examples - I like it because it allows easy to add more examples).

Comment (evgeniy-r) [reply tests/sample_files.rs]: Done

**Comment by mgrachev** [comment]
@evgeniy-r Excellent work! Thank you! 🔥
